### PR TITLE
feat(serde): add support for Serde serialization

### LINKED ISSUE #33: Support Serde
Needed for https://github.com/nickel-org/nickel.rs/issues/345

It'd be good to support `serde` in templates, it doesn't look like this crate has any serialisable structs of its own, so it shouldn't be necessary to add dependencies on `syntex`.

### PR DESCRIPTION
**Pull request by Ryman**
To use this feature, you can activate the serde feature:

``` toml
[dependencies.mustache]
version = "0.8"
features = ["serde"]
```

The minimum compiler version for the serde feature will track stable.

Closes #33.

### CODE REVIEW

**Comment by Ryman** [comment]
r?  @KodrAus 

Specific review concerns: 
- I'm sure there's probably neater ways to do some of the `cfg` stuff
- I'm unsure if re-exporting both traits as `Encodable` makes any sense but it simplifies the imports and usage within this crate so hopefully it's not confusing for users?

Still TODO if I get to it: more tests for serialization. This passes what we currently have but it's could be more exhaustive on our 'compatible types'. Any specific edge case suggestions would definitely be appreciated!

File: src/encoder/serialization/serde.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            Some(OptVal(_)) => return Err(NestedOptions),
+            Some(d) => d,
+            _ => {
+                return Err(UnsupportedType);
+            }
+        };
+        self.push_ok(OptVal(Some(Box::new(val))))
+    }
+
+    fn serialize_seq<V>(&mut self, mut visitor: V) -> EncoderResult
```
Comment (KodrAus) [inline src/encoder/serialization/serde.rs]: The way `Serializer` deals with seqs and maps is changing (in master, but not crates.io). Check out: https://github.com/serde-rs/serde/pull/437

Instead of using visitors, you have a single associated type for handling the state of a map being serialised.

File: examples/vector_iteration.rs [hunk 2]
```diff
@@ -1,14 +1,28 @@
...
 use std::collections::HashMap;
 
-#[derive(RustcEncodable)]
+#[cfg_attr(not(feature = "serde"), derive(RustcEncodable))]
 struct User {
     name: String
 }
 
+#[cfg(feature = "serde")]
+impl serde::ser::Serialize for User {
```
Comment (KodrAus) [inline examples/vector_iteration.rs]: Since you've got a build script, you could `#[derive(Serialize)]` for `User`. Or is there a reason you're using a map instead?

File: examples/vector_iteration.rs [hunk 3]
```diff
@@ -1,14 +1,28 @@
...
+    fn serialize<S>(&self, serializer: &mut S) -> Result<(), S::Error>
+    where S: serde::ser::Serializer {
+        let mut map = HashMap::new();
```
Comment (KodrAus) [inline examples/vector_iteration.rs]: Should this be `HashMap::with_capacity(1)`?

**Comment by KodrAus** [comment]
I think re-exporting both traits as the same name makes sense, since other libraries that only want to maintain one or the other can still satisfy the type bound by implementing `rustc_serialize::Encodable` or `serde::Serialize` (I'm pretty sure).

I'll pull this down to my `nickel` branch and see how it all plays :) Looks good to me though!

File: examples/vector_iteration.rs (hunk 2 above)
Comment (Ryman) [inline examples/vector_iteration.rs]: I think having to `include!` a code-genned file in example code is a bit confusing for people not using serde. (if someone already *is* interested in using serde they would find better ways to do this in serde's docs!)

Comment (KodrAus) [inline examples/vector_iteration.rs]: Ah of course, sorry I missed the `examples` path! I understand why you're doing it this way now.

Comment (Ryman) [inline examples/vector_iteration.rs]: I think you deleted the other comment or else my github is broken, but I wrote a little bit about the hashmap behavior (or why not bother with `with_capacity(1)`). It can be summarised with a [short gist](https://play.rust-lang.org/?gist=9d7de03200c32313d75fa702bd57a871&version=stable&backtrace=0), I found it fun to look into the guts to ensure my trust in rust's defaults were sane!

File: src/encoder/serialization/serde.rs (hunk 1 above)
Comment (Ryman) [inline src/encoder/serialization/serde.rs]: Yep, something to look forward to! It should be fine to continue with this though as if they break the api they should be bumping to `0.8`.

File: examples/vector_iteration.rs (hunk 2 above)
Comment (KodrAus) [inline examples/vector_iteration.rs]: Yeh, I did actually delete it because I'd made that comment thinking it might affect performance before you pointed out it was actually an example file :) Thanks for the gist! I was under the impression that if you set a capacity explicitly it wouldn't resize until it fills up, like `Vec`.

File: Cargo.toml [hunk 4]
```diff
@@ -7,12 +7,20 @@ version = "0.7.0"
...
+build = "build.rs"
+
 [features]
 unstable = []
+test-serde = ["serde", "serde_json", "serde_codegen"]
 
 [dependencies]
 rustc-serialize = "0.3.18"
 log = "0.3.5"
+serde = { version = "0.7", optional = true }
```
Comment (KodrAus) [inline Cargo.toml]: `serde 0.8` is now officially a thing. Can we upgrade to it? I'm happy to help out if it turns out to be a lot of effort.

**Comment by dtolnay** [comment]
What is the status of this? Waiting on anything?

**Comment by KodrAus** [comment]
Well this would need to be updated for `serde 0.9`. Looking at this again I was going to test this branch in my `nickel` PR. I did end up doing that, and it all worked out well. I didn't come back and report that though, so I think I dropped the ball.

I'd be happy to see it start moving again though.

**Comment by petehayes102** [comment]
It'd be great to have some movement on this. Especially considering `rustc-serialize` is planned to be deprecated in 1.18: https://internals.rust-lang.org/t/rust-release-milestone-predictions/4591

**Comment by erickt** [comment]
FYI, I've been working on updating this to work with serde 1.0.

**Comment by Ryman** [comment]
Closing in favor of #52
